dali/pipeline/data/buffer.h

```diff
@@ -203,107 +211,81 @@ class Buffer {
   inline void set_type(const TypeInfo &new_type) {
     DALI_ENFORCE(IsValidType(new_type), "new_type must be valid type.");
     if (new_type == type_) return;
-
-    if (!IsValidType(type_)) {
-      // If the buffer has no type, set the type to the
-      // calling type and allocate the buffer
-      DALI_ENFORCE((data_ == nullptr) || shares_data_,
-          "Buffer has no type and does not share data, "
-          "data_ should be nullptr.");
-      DALI_ENFORCE((num_bytes_ == 0) || shares_data_,
-          "Buffer has no type and does not share data, "
-          "num_bytes_ should be 0.");
-    }
     type_ = new_type;
 
     size_t new_num_bytes = size_ * type_.size();
     if (new_num_bytes > num_bytes_) {
       new_num_bytes *= alloc_mult;
+      reserve(new_num_bytes);
+    }
+  }
+
+  inline void reserve(size_t new_num_bytes) {
+    if (new_num_bytes <= num_bytes_)
+      return;
 
-      // re-allocating: get the device
-      if (std::is_same<Backend, GPUBackend>::value) {
-        CUDA_CALL(cudaGetDevice(&device_));
-      }
-      data_.reset(Backend::New(new_num_bytes, pinned_), std::bind(
-              &Buffer<Backend>::DeleterHelper,
-              this, std::placeholders::_1,
-              type_, size_, device_, pinned_));
-      num_bytes_ = new_num_bytes;
-      shares_data_ = false;
+    // re-allocating: get the device
+    if (std::is_same<Backend, GPUBackend>::value) {
+      CUDA_CALL(cudaGetDevice(&device_));
     }
 
-    type_.template Construct<Backend>(data_.get(), size_);
+    data_.reset(
+      Backend::New(new_num_bytes, pinned_),
+      std::bind(FreeMemory, std::placeholders::_1, new_num_bytes, device_, pinned_));
+
+    num_bytes_ = new_num_bytes;
+
+    // If we were sharing data, we aren't anymore
+    shares_data_ = false;
+  }
+
+  inline void free() {
+    if (!shares_data_) {
```

Comment:
I think it is safe to write:
```
data_.reset();
num_bytes_ = 0;
shares_data_ = false;
```
In case of shares_data_ == true, data_ deleter is empty anyway.